File: src/pathoverse/api/schemas/analysis.py
from typing import Optional

from pydantic import BaseModel, Field


class ClassificationRequest(BaseModel):
    model: str
    dataset: str = "patchcamelyon"


class ClassificationResponse(BaseModel):
    model: str
    dataset: str
    accuracy: Optional[float] = None
    auroc: Optional[float] = None
    f1: Optional[float] = None
    precision: Optional[float] = None
    recall: Optional[float] = None
    sensitivity: Optional[float] = None
    specificity: Optional[float] = None


class MILRequest(BaseModel):
    slide_id: str
    model: str = "gigapath-flash"


class MILFoundationModel(BaseModel):
    name: str
    embedding_dimension: int


class MILMetadata(BaseModel):
    architecture: str
    hidden_dimension: int
    attention_dimension: int
    num_classes: int
    trained: bool
    prototype: bool
    seed: Optional[int] = None


class MILPrediction(BaseModel):
    class_id: int
    probability: float


class MILAttention(BaseModel):
    tile_count: int
    sum: float
    top_k: int
    top_tiles: list[dict] = Field(
        default_factory=list
    )


class MILResponse(BaseModel):
    schema_version: str
    task: str
    slide_id: str
    model: str

    foundation_model: MILFoundationModel
    mil: MILMetadata
    prediction: MILPrediction
    attention: MILAttention

    slide_embedding_dimension: int
    slide_embedding_path: str

    trained: bool
    prototype: bool

File: frontend/src/types/analysis.ts
export interface ModelInfo {
  name: string;
  model_id: string;
  embedding_dimension: number;
  input_size: number;
  modality: string;
  description: string;
  status: string;
}

export interface SlideInfo {
  slide_id: string;
  filename: string;
  width: number;
  height: number;
  levels: number;
  mpp_x: number;
  mpp_y: number;
  tissue_ratio: number | null;
  tile_count: number | null;
}

export interface ClassificationResponse {
  model: string;
  dataset: string;
  accuracy: number | null;
  auroc: number | null;
  f1: number | null;
  precision: number | null;
  recall: number | null;
  sensitivity: number | null;
  specificity: number | null;
}

export interface MILFoundationModel {
  name: string;
  embedding_dimension: number;
}

export interface MILMetadata {
  architecture: string;
  hidden_dimension: number;
  attention_dimension: number;
  num_classes: number;
  trained: boolean;
  prototype: boolean;
  seed: number | null;
}

export interface MILPrediction {
  class_id: number;
  probability: number;
}

export interface MILTopTile {
  rank: number;
  tile_index: number;
  tile_id: number;
  attention: number;
  x: number;
  y: number;
  width: number;
  height: number;
  level: number;
  tissue_ratio: number;
  image_path?: string;
}

export interface MILAttention {
  tile_count: number;
  sum: number;
  top_k: number;
  top_tiles: MILTopTile[];
}

export interface MILResponse {
  schema_version: string;
  task: string;
  slide_id: string;
  model: string;
  foundation_model: MILFoundationModel;
  mil: MILMetadata;
  prediction: MILPrediction;
  attention: MILAttention;
  slide_embedding_dimension: number;
  slide_embedding_path: string;
  trained: boolean;
  prototype: boolean;
}